Add dict page

kangpinghuang opened the pull request:
Add dict encoding page for binary/string type data. Construct a dict for original data, and save encoded id instead of origin data to save space. If the dict is too big, then will automatically fall
back to plain encoding.

gaodayue commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+    _encoding_type = static_cast<EncodingTypePB>(type);
+    Slice content(_data.data + 4, _data.size - 4);
+    if (_encoding_type == DICT_ENCODING) {
+        _data_page_decoder.reset(new BitShufflePageDecoder<OLAP_FIELD_TYPE_INT>(content, _options));
```
check options.dict_decoder is not null

gaodayue commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+    }
+    size_t type = decode_fixed32_le((const uint8_t*)&_data.data[0]);
+    _encoding_type = static_cast<EncodingTypePB>(type);
+    Slice content(_data.data + 4, _data.size - 4);
```
```suggestion
    _data.remove_prefix(BINARY_DICT_PAGE_HEADER_SIZE);
```

gaodayue commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+    } else {
+        // dictionary encoding
+        DCHECK(_parsed);
+        DCHECK(_dict_decoder != nullptr) << "dict decoder pointer is nullptr";
```
move this check to init()

kangpinghuang replied:
ok

gaodayue commented on be/src/olap/rowset/segment_v2/encoding_info.h:
```diff
@@ -31,8 +34,6 @@ namespace segment_v2 {
 
 class PageBuilder;
 class PageDecoder;
```
forward declaration of PageBuilder and PageDecoder could also be removed since we've included the header

gaodayue reviewed (changes requested):
According to https://stackoverflow.com/questions/11162267/how-do-i-make-git-treat-a-file-as-binary , you could rename `dict_encoding_data` to `dict_encoding_data.dat` and use .gitattributes file to mark test data file as binary file

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+                }
+                uint32_t value_code = _dictionary.size();
+                _dictionary.insert({current_item, value_code});
+                add_count = 1;
```
Ok, you are right

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+                RETURN_IF_ERROR(_dict_builder.add(reinterpret_cast<const uint8_t*>(src), &add_count));
+                if (add_count == 0) {
+                    // the dict page is full, so just return
+                    *count = num_added;
```
ok, here just break

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+            if (ret != _dictionary.end()) {
+                uint32_t value_code = ret->second;
+                RETURN_IF_ERROR(_data_page_builder->add(reinterpret_cast<const uint8_t*>(&value_code), &add_count));
+                num_added += add_count;
```
ok

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+                RETURN_IF_ERROR(_data_page_builder->add(reinterpret_cast<const uint8_t*>(&value_code), &add_count));
+                if (add_count == 0) {
+                    // the data page is full, just return
+                    *count = num_added;
```
ok

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+                    *count = num_added;
+                    return Status::OK();
+                }
+                num_added += add_count;
```
ok

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,213 @@
+
+    if (_encoding_type == DICT_ENCODING
+            && _dict_builder.is_page_full()) {
+        // TODO(hkp): set to binary plain page builder
```
ok

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.h:
```diff
@@ -0,0 +1,116 @@
+    BinaryPlainPageBuilder _dict_builder;
+
+    EncodingTypePB _encoding_type;
+    std::unordered_map<std::string, uint32_t> _dictionary;
```
en, dictionary is stored twice, but I think the method you recommend is wrong because the faststring may relocated and the slice may be wrong.

kangpinghuang replied:
I will delay the dictionary encoding to get_dictionary_page.

morningman reviewed (changes requested):
Please add comment in commit log to describe what is dict page and what it is for

gaodayue commented on be/test/olap/rowset/segment_v2/binary_dict_page_test.cpp:
```diff
@@ -0,0 +1,230 @@
+    std::ifstream infile(file.c_str());
+    while(getline(infile, line)) {
+        // just for ut
+        std::string* dumped = new std::string(line);
```
resource leaked. why not just use `std::vector<std::string> src_strings;`?

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,234 @@
+    _dict_builder(nullptr),
+    _encoding_type(DICT_ENCODING) {
+    // initially use DICT_ENCODING
+    _data_page_builder.reset(new BitshufflePageBuilder<OLAP_FIELD_TYPE_INT>(options));
```
This will be done by add a PageBuilderFactory in the future? I will add a TODO here

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,234 @@
+                    return Status::Corruption(Substitute("memory allocate failed, size:$0", src->size));
+                }
+                memcpy(item_mem, src->data, src->size);
+                Slice dict_item((char*)item_mem, src->size);
```
I think the relocate of slice can be used here

Commit:
Add dict page

Dict Page can be used to encoding string/binary type column.
A dict will be constructed during encoding, and the encoded
id will be saved to replace the origin data to increase
compression ratio.

Commit:
Fix compile problems

There is compile problems because rebase to master.
Modify some interface to fix compile problems

gaodayue commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,231 @@
+    }
+    *dictionary_page = _dict_builder->finish();
+    _dict_builder->release();
+    _dict_items.clear();
```
at this point, we can delete BinaryPlainPageBuilder by `_dict_builder.reset()` to free some memory

kangpinghuang replied:
it is in line 136

gaodayue commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,231 @@
+        // copy the codewords into a temporary buffer first
+        // And then copy the strings corresponding to the codewords to the destination buffer
+        BitShufflePageDecoder<OLAP_FIELD_TYPE_INT>* data_ptr =
+                down_cast<BitShufflePageDecoder<OLAP_FIELD_TYPE_INT>*>(_data_page_decoder.get());
```
Why do we need the raw pointer here? Couldn't we simply use _data_page_decoder->next_batch in the following code?

kangpinghuang replied:
my mistake. I will change it.

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,231 @@
+    _finished = false;
+    _buffer.clear();
+    _buffer.resize(BINARY_DICT_PAGE_HEADER_SIZE);
+    _buffer.reserve(_options.data_page_size);
```
ok

Commit:
Fix pr problem

1. use base pointer to next_batch
2. optimize buffer usage

Commit:
delete duplicated code

gaodayue commented on be/src/olap/rowset/segment_v2/bitshuffle_page.h:
```diff
@@ -170,11 +148,11 @@ class BitshufflePageBuilder : public PageBuilder {
 
         encode_fixed32_le(&_buffer[0], _count);
-        int64_t bytes = bitshuffle::compress_lz4(_data.data(), &_buffer[BITSHUFFLE_BLOCK_HEADER_SIZE],
+        uint32_t bytes = bitshuffle::compress_lz4(_data.data(), &_buffer[BITSHUFFLE_PAGE_HEADER_SIZE],
```
please fix this also

Commit:
Modify return type of compress_lz4

imay commented on be/src/olap/rowset/segment_v2/options.h:
```diff
@@ -17,20 +17,21 @@
 
 struct PageDecoderOptions {
-    PageDecoder* dict_decoder = nullptr;
+    const BinaryPlainPageDecoder* dict_decoder = nullptr;
```
I think shared_ptr is better.

kangpinghuang replied:
ok

imay commented on be/src/olap/rowset/segment_v2/encoding_info.h:
```diff
@@ -22,18 +22,16 @@
 
 namespace segment_v2 {
 
-class PageBuilder;
```
why change forward declaration to include?

kangpinghuang replied:
I will fall back to forward declaration

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,223 @@
+        size_t num_added = 0;
+        uint32_t value_code = -1;
+        for (int i = 0; i < *count; ++i, ++src) {
+            auto ret = _dictionary.find(*src);
```
```suggestion
            auto iter = _dictionary.find(*src);
```

kangpinghuang replied:
ok

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,223 @@
+                }
+                char* item_mem = _arena.Allocate(src->size);
+                if (item_mem == nullptr) {
+                    return Status::Corruption(Substitute("memory allocate failed, size:$0", src->size));
```
it is not corruption error, MemoryAllocFailed

kangpinghuang replied:
ok

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,223 @@
+        uint32_t value_code = -1;
+        for (int i = 0; i < *count; ++i, ++src) {
+            auto ret = _dictionary.find(*src);
+            size_t add_count = 1;
```
define it just before you use it

kangpinghuang replied:
ok

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,223 @@
+                Slice dict_item(src->data, src->size);
+                dict_item.relocate(item_mem);
+                value_code = _dictionary.size();
+                _dictionary.insert({dict_item, value_code});
```
```suggestion
                _dictionary.emplace(dict_item, value_code);
```

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,223 @@
+Status BinaryDictPageDecoder::init() {
+    CHECK(!_parsed);
+    if (_data.size < BINARY_DICT_PAGE_HEADER_SIZE) {
+        LOG(WARNING) << "corrupted data, data size:" << _data.size;
```
I think it's better not to print log here, let client print log

kangpinghuang replied:
ok

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,223 @@
+        return Status::Corruption(Substitute("invalid encoding type:$0", _encoding_type));
+    }
+
+    Status status = _data_page_decoder->init();
```
RETURN_IF_ERROR

kangpinghuang replied:
ok

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,223 @@
+Status BinaryDictPageDecoder::next_batch(size_t* n, ColumnBlockView* dst) {
+    if (_encoding_type == PLAIN_ENCODING) {
+        return _data_page_decoder->next_batch(n, dst);
+    } else {
```
remove else?

kangpinghuang replied:
No, 'else' here process data page is encoded by bitshuffle case. It should change the dict code to real string value here.

imay commented on be/src/olap/rowset/segment_v2/binary_dict_page.cpp:
```diff
@@ -0,0 +1,223 @@
+            Slice element = _dict_decoder->string_at_index(codeword);
+            char* destination = dst->column_block()->arena()->Allocate(element.size);
+            if (destination == nullptr) {
+                return Status::Corruption(Substitute("memory allocate failed, size:$0", element.size));
```
MemoryAllocFailed

kangpinghuang replied:
ok

Commit:
Fix pr problems

1. change status Corruption to MemoryAllocFailed
2. remove invalid log
3. replace insert with emplace
4. revert include to forward declaration

Commit:
Remove extra space

Commit:
remove else

morningman reviewed (approved):
LGTM
